core/oracles/ethereum_oracle_verifier.go

```diff
@@ -0,0 +1,268 @@
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.hashes[hash]; ok {
+		return false
+	}
+
+	s.hashes[hash] = struct{}{}
+
+	return true
+}
+
+// TODO: non finalized events could cause a memory leak, this needs to be addressed in a way that will prevent processing
+// duplicates but not result in a memory leak, agreed to postpone for now  (other verifiers have the same issue)
+
+func (s *EthereumOracleVerifier) ProcessEthereumContractCallResult(callEvent types.EthContractCallEvent) error {
+	// Possible that a single oracle trigger, where trigger is base off of vegatime, could result in chain events with
+	// different contents (and therefore different hashes) due to difference in the time when the validators query
+	// ethereum, it is assumed that the consumer of the oracle events will handle this.  Alternatively we could restrict
+	// triggers to use ethereum time and not allow vega time triggers, then no such issue.
+	if ok := s.ensureNotDuplicate(callEvent.Hash()); !ok {
+		s.log.Error("ethereum call event already exists",
+			logging.String("event", fmt.Sprintf("%+v", callEvent)))
+		return ErrDuplicatedEthereumCallEvent
+	}
+
+	pending := &pendingCallEvents{
+		callEvent: callEvent,
+		check:     func() error { return s.checkCallEventResult(callEvent) },
+	}
+
+	s.pendingCallEvents = append(s.pendingCallEvents, pending)
+
+	s.log.Info("ethereum call event received, starting validation",
+		logging.String("call-event", fmt.Sprintf("%+v", callEvent)))
+
+	// Timeout for the check set to 1 day, to allow for validator outage scenarios
+	return s.witness.StartCheck(
+		pending, s.onCallEventVerified, s.timeService.GetTimeNow().Add(24*time.Hour))
+}
+
+func (s *EthereumOracleVerifier) checkCallEventResult(contractCall types.EthContractCallEvent) error {
+	spec, err := s.ethCallSpecSource.GetSpec(contractCall.SpecId)
+	if err != nil {
+		return fmt.Errorf("failed to get call specification for id %s: %w", contractCall.SpecId, err)
+	}
+
+	blockHeight := &big.Int{}
+	blockHeight.SetUint64(contractCall.BlockHeight)
+	value, err := spec.Call(context.Background(), s.ethContractCaller, blockHeight)
+	if err != nil {
+		return fmt.Errorf("failed to execute call event spec: %w", err)
+	}
+
+	if !bytes.Equal(contractCall.Result, value) {
+		return fmt.Errorf("mismatched results for block %d", contractCall.BlockHeight)
+	}
+
+	if !spec.PassesFilters(contractCall.Result, contractCall.BlockHeight, contractCall.BlockTime) {
+		return fmt.Errorf("failed to pass filter check")
+	}
+
+	if err = s.ethConfirmations.Check(spec.RequiredConfirmations()); err != nil {
+		return fmt.Errorf("failed confirmations check: %w", err)
+	}
+
+	return nil
+}
+
+func (s *EthereumOracleVerifier) removePendingCallEvent(id string) error {
+	for i, v := range s.pendingCallEvents {
+		if v.GetID() == id {
+			s.pendingCallEvents = s.pendingCallEvents[:i+copy(s.pendingCallEvents[i:], s.pendingCallEvents[i+1:])]
```

Comment:
Not sure if we want to be iterating over a slice while also removing things from it. Does `pendingCallEvents` need to be ordered? if it was a map instead and then this is just `delete(s.pendingCallEvents, id)`.